### PR TITLE
Use port in no_os_gpio_init_param in STM32 platforms

### PR DESCRIPTION
**Pull request by danmois**
Use port in no_os_gpio_init_param in GPIO initialization for STM32 platforms and remove port from stm32_gpio_init_param.

### CODE REVIEW

**Review by amiclaus — COMMENTED** [review]
there are some CI checks that are failing. also, are there projects that need update after these changes?

File: drivers/platform/stm32/stm32_gpio.c [hunk 1]
```diff
@@ -277,7 +278,6 @@ int32_t stm32_gpio_direction_output(struct no_os_gpio_desc *desc,
 	HAL_GPIO_WritePin(extra->port, NO_OS_BIT(desc->number), (GPIO_PinState)value);
 
 	/* configure gpio with user configuration */
-	extra->gpio_config.Mode = GPIO_MODE_OUTPUT_PP;
```
Comment (buha) [inline drivers/platform/stm32/stm32_gpio.c:264]: I think you left some things out, like this one:
https://github.com/analogdevicesinc/no-OS/blob/master/drivers/platform/stm32/stm32_gpio.c#L285

File: drivers/platform/stm32/stm32_gpio.c [hunk 2]
```diff
@@ -112,6 +112,15 @@ static int32_t _gpio_init(struct no_os_gpio_desc *desc,
 	else
 		return -EINVAL;
 
+	if (!param->extra) {
+		pextra = (struct stm32_gpio_init_param*)malloc(sizeof(*pextra));
```
Comment (buha) [inline drivers/platform/stm32/stm32_gpio.c:116]: this is not ok. desc->extra needs to exist, it is used in all other API functions

better drop this commit and leave the stm32_init_param stuff as mandatory.

File: drivers/platform/stm32/stm32_gpio.c (hunk 1 above)
Comment (danmois) [reply drivers/platform/stm32/stm32_gpio.c:264]: The idea here is that _gpio_init has already set desc->extra to the initialization parameters (either given by application or default values) and here we just call init once again.

File: drivers/platform/stm32/stm32_gpio.c (hunk 2 above)
Comment (danmois) [reply drivers/platform/stm32/stm32_gpio.c:116]: desc->extra exists, the only thing that could not exist is stm32_init_param.

File: drivers/platform/stm32/stm32_gpio.c [hunk 3]
```diff
@@ -135,6 +161,10 @@ static int32_t _gpio_init(struct no_os_gpio_desc *desc,
 
 	HAL_GPIO_Init(extra->port, &extra->gpio_config);
 
+	if (!param->extra) {
+		free(pextra);
```
Comment (CiprianRegus) [inline drivers/platform/stm32/stm32_gpio.c:165]: `pextra` is a NULL pointer, no need to free it.

**Comment by danmois** [comment]
V2
- remove use of pextra if extra does not exist in stm32 gpio init params

**Review by buha — APPROVED** [review]

File: drivers/platform/stm32/stm32_gpio.c [hunk 4]
```diff
@@ -282,8 +308,8 @@ int32_t stm32_gpio_get_direction(struct no_os_gpio_desc *desc,
 		return -EINVAL;
 
 	struct stm32_gpio_desc *extra = desc->extra;
-	*direction = (extra->gpio_config.Mode & GPIO_MODE_OUTPUT_PP) ? NO_OS_GPIO_OUT :
-		     NO_OS_GPIO_IN;
+	*direction = (extra->gpio_config.Mode & GPIO_MODE_INPUT) ? NO_OS_GPIO_IN :
```
[the remark this answers is not in the document]
Comment (danmois) [reply drivers/platform/stm32/stm32_gpio.c:311]: You are right, will change to if-else statement.

**Comment by danmois** [comment]
V4

- rewrite the logic for determining GPIO direction in stm32_gpio_get_direction
- simplfy logic in stm32_gpio_direction_output for better readability

File: projects/adxrs290-pmdz/src/platform/stm32/parameters.c [hunk 5]
```diff
@@ -51,6 +51,5 @@ struct stm32_spi_init_param adxrs290_spi_extra_ip  = {
 
 struct stm32_gpio_init_param adxrs290_gpio_extra_ip = {
 	.mode = GPIO_MODE_INPUT,
-	.port = GPIOA,
```
Comment (buha) [inline projects/adxrs290-pmdz/src/platform/stm32/parameters.c:54]: In adxrs290 project, please add .port = 0 explicitly.
It's very weird to select GPIOA implicitly, without initializing .port.

File: drivers/platform/stm32/stm32_spi.c [hunk 6]
```diff
@@ -186,7 +186,6 @@ int32_t stm32_spi_init(struct no_os_spi_desc **desc,
 	spi_desc->extra = sdesc;
 	sinit = param->extra;
 
-	csip_extra.port = sinit->chip_select_port;
```
Comment (buha) [inline drivers/platform/stm32/stm32_spi.c:189]: Also this change does not belong here but to another commit.

File: projects/adxrs290-pmdz/src/platform/stm32/parameters.c (hunk 5 above)
Comment (danmois) [reply projects/adxrs290-pmdz/src/platform/stm32/parameters.c:54]: GPIOA is selected in parameters.h (https://github.com/analogdevicesinc/no-OS/blob/18d5dc15506bd3c41eec5f030a22039bdac6a264/projects/adxrs290-pmdz/src/platform/stm32/parameters.h#L73) as a member of adxrs290_gpio_sync_ip (https://github.com/analogdevicesinc/no-OS/blob/96f4a44948128aecf687343973d0e53470da0c6d/projects/adxrs290-pmdz/src/common/common_data.c#L59).

File: drivers/platform/stm32/stm32_spi.c (hunk 6 above)
Comment (danmois) [reply drivers/platform/stm32/stm32_spi.c:189]: Split last commit.

**Comment by danmois** [comment]
V5

- split last commit

File: projects/adxrs290-pmdz/src/platform/stm32/parameters.c (hunk 5 above)
Comment (buha) [reply projects/adxrs290-pmdz/src/platform/stm32/parameters.c:54]: ok, i missed it

Comment (danmois) [reply projects/adxrs290-pmdz/src/platform/stm32/parameters.c:54]: Will also set SPI_CS_PORT pins correctly

**Comment by danmois** [comment]
V6

- SPI_CS_PORT update

**Review by buha — APPROVED** [review]

**Comment by buha** [comment]
looks like we can merge this
